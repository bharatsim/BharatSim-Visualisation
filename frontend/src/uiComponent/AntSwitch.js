import withStyles from '@material-ui/core/styles/withStyles';
import Switch from '@material-ui/core/Switch';
import React from 'react';
import Box from '@material-ui/core/Box';
import { makeStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import { useController } from 'react-hook-form';

const switchContainerStyles = makeStyles({
  container: {
    display: 'flex',
  },
});

const AntSwitchWithStyle = withStyles((theme) => ({
  root: {
    width: 28,
    height: 16,
    padding: 0,
    display: 'flex',
  },
  switchBase: {
    padding: 2,
    color: theme.palette.grey[500],
    '&$checked': {
      transform: 'translateX(12px)',
      color: theme.palette.common.white,
      '& + $track': {
        opacity: 1,
        backgroundColor: theme.colors.primaryColorScale['500'],
        borderColor: theme.colors.primaryColorScale['500'],
      },
    },
  },
  thumb: {
    width: 12,
    height: 12,
    boxShadow: 'none',
  },
  track: {
    border: `1px solid ${theme.palette.grey[500]}`,
    borderRadius: 16 / 2,
    opacity: 1,
    backgroundColor: theme.palette.common.white,
  },
  checked: {},
}))(Switch);

function AntSwitch({ onLabel, offLabel, name, control, dataTestid }) {
  const switchContainerClasses = switchContainerStyles();
  const {
    field: { ref, onChange, value },
  } = useController({ name, control, defaultValue: false });

  return (
    <Box className={switchContainerClasses.container}>
      <Box mr={1}>{offLabel}</Box>
      <AntSwitchWithStyle
        inputRef={ref}
        onChange={(e) => onChange(e.target.checked)}
        checked={value}
        data-testid={dataTestid}
      />
      <Box ml={1}>{onLabel}</Box>
    </Box>
  );
}

AntSwitch.propTypes = {
  onLabel: PropTypes.string.isRequired,
  offLabel: PropTypes.string.isRequired,
  dataTestid: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  control: PropTypes.shape({}).isRequired,
};
export default AntSwitch;
