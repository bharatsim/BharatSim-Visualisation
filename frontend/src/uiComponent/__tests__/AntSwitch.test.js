import React from 'react';
import { act, render } from '@testing-library/react';
import { useForm } from 'react-hook-form';
import { fireEvent } from '@testing-library/dom';
import AntSwitch from '../AntSwitch';

const FormWithAntSwitch = ({ onSubmit }) => {
  const { control, handleSubmit } = useForm({ mode: 'onChange' });
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <AntSwitch control={control} name="switch" dataTestid="switch" offLabel="on" onLabel="off" />
      <button type="submit">submit</button>
    </form>
  );
};

describe('<AntSwitch />', () => {
  it('should call on submit with ant switch value', async () => {
    const onSubmit = jest.fn();
    const { getByRole, getByText } = render(<FormWithAntSwitch onSubmit={onSubmit} />);

    getByRole('checkbox').click();
    fireEvent.change(getByRole('checkbox'), { target: { checked: true } });

    await act(async () => {
      fireEvent.click(getByText('submit'));
    });

    expect(onSubmit).toHaveBeenCalledWith(
      {
        switch: true,
      },
      expect.anything(),
    );
  });
});
