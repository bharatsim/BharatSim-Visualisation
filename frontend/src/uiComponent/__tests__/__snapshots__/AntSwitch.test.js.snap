// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AntSwitch /> should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-2 makeStyles-container-1"
  >
    <div
      class="MuiBox-root MuiBox-root-3"
    >
      OFF
    </div>
    <span
      class="MuiSwitch-root WithStyles(ForwardRef(Switch))-root-4"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiSwitch-switchBase WithStyles(ForwardRef(Switch))-switchBase-5 MuiSwitch-colorSecondary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-12 MuiSwitch-input"
            type="checkbox"
            value=""
          />
          <span
            class="MuiSwitch-thumb WithStyles(ForwardRef(Switch))-thumb-6"
          />
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track WithStyles(ForwardRef(Switch))-track-7"
      />
    </span>
    <div
      class="MuiBox-root MuiBox-root-13"
    >
      ON
    </div>
  </div>
</div>
`;
